examples/step-69/step-69.cc

```diff
@@ -0,0 +1,2893 @@
+    // \mathbb{R}^d$. If $d=2$ then $\mathbf{c}_{ij} =
+    // [\mathbf{c}_{ij}^1,\mathbf{c}_{ij}^2]^\top$. Which basically implies
+    // that we need one matrix per space dimension to store the
+    // $\mathbf{c}_{ij}$ vectors. Similar observation follows for the
+    // matrix $\mathbf{n}_{ij}$. The purpose of
+    // <code>gather_get_entry()</code> is to retrieve those entries a store
+    // them into a <code>Tensor<1, dim></code> for our convenience.
+
+    template <typename T1, std::size_t k, typename T2>
+    DEAL_II_ALWAYS_INLINE inline Tensor<1, k>
+    gather_get_entry(const std::array<T1, k> &U, const T2 it)
+    {
+      Tensor<1, k> result;
+      for (unsigned int j = 0; j < k; ++j)
+        result[j] = get_entry(U[j], it);
+      return result;
+    }
+
+    // <code>gather()</code> (first interface): this first function
+    // signature, having three input arguments, will be used to retrieve
+    // the individual components <code>(i,l)</code> of a matrix. The
+    // functionality of <code>gather_get_entry()</code> and
+    // <code>gather()</code> is very much the same, but their context is
+    // different: the function <code>gather()</code> does not rely on an
+    // iterator (that actually knows the value pointed) but rather on the
+    // indices <code>(i,l)</code> of the entry in order to retrieve its
+    // actual value. We should expect <code>gather()</code> to be slightly
+    // more expensive than <code>gather_get_entry()</code>. The use of
+    // <code>gather()</code> will be limited to the task of computing the
+    // algebraic viscosity $d_{ij}$ in the particular case that when
+    // both $i$ and $j$ lie at the boundary.
+    //
+    // @note The reader should be aware that accessing an arbitrary
+    // <code>(i,l)</code> entry of a matrix (say for instance Trilinos or PETSc
+    // matrices) is in general unacceptably expensive. Here is where we might
+    // want to keep an eye on complexity: we want this operation to have
+    // constant complexity, which is the case of the current implementation
+    // using deal.ii matrices.
+
+    template <typename T1, std::size_t k, typename T2, typename T3>
+    DEAL_II_ALWAYS_INLINE inline Tensor<1, k>
+    gather(const std::array<T1, k> &U, const T2 i, const T3 l)
+    {
+      Tensor<1, k> result;
+      for (unsigned int j = 0; j < k; ++j)
+        result[j] = U[j](i, l);
+      return result;
+    }
+
+    // <code>gather()</code> (second interface): this second function
+    // signature having two input arguments will be used to gather the
+    // state at a node <code>i</code> and return it as a
+    // <code>Tensor<1,problem_dimension></code> for our convenience.
+
+    template <typename T1, std::size_t k, typename T2>
+    DEAL_II_ALWAYS_INLINE inline Tensor<1, k> gather(const std::array<T1, k> &U,
+                                                     const T2                 i)
```

Comment:
same as above

Reply:
I suspect that `U` here is an array of vectors, not an array of matrices? My confusion suggests that using the same `T1` as above might just be too terse a naming choice :-)